pkg/operator/terminationobserver/termination_observer.go

```diff
@@ -0,0 +1,200 @@
+		"TerminationPreShutdownHooksFinished",
+		"TerminationPreShutdownHooksFinished",
+		"TerminationMinimalShutdownDurationFinished",
+		"TerminationStoppedServing",
+		"TerminationGracefulTerminationFinished",
+	}
+)
+
+// TerminationObserver observes static pods that are terminating. When the API server static pod is replaced by
+// new revision or the pod is evicted or removed, the static pods are not reporting the terminating state back
+// to API server, but they only change the creationTimestamp.
+// We need to capture the termination events produced by the pods that we no longer see.
+type TerminationObserver struct {
+	targetNamespace string
+
+	podsGetter corev1client.PodsGetter
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+
+	store Store
+}
+
+func NewTerminationObserver(
+	targetNamespace string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	podsGetter corev1client.PodsGetter,
+	store Store,
+	eventRecorder events.Recorder,
+) *TerminationObserver {
+	c := &TerminationObserver{
+		targetNamespace: targetNamespace,
+		podsGetter:      podsGetter,
+		eventRecorder:   eventRecorder.WithComponentSuffix("termination-observer"),
+
+		store: store,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TerminationObserver"),
+	}
+
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Events().Informer().AddEventHandler(c.terminationEventRecorder())
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Events().Informer().HasSynced)
+
+	return c
+}
+
+func (c *TerminationObserver) sync() error {
+	podList, err := c.podsGetter.Pods(c.targetNamespace).List(metav1.ListOptions{LabelSelector: "app=openshift-kube-apiserver"})
+	if err != nil {
+		return fmt.Errorf("unable to list pods in %q namespace: %v", c.targetNamespace, err)
+	}
+
+	for _, pod := range podList.Items {
+		// get the last seen creation timestamp for this api server
+		currentServerState, err := c.store.Get(pod.Name)
+
+		// we haven't seen this apiserver yet, just record the creation timestamp and move to next
+		if err == ApiServerNotFoundError || currentServerState == nil {
+			c.store.Create(pod.Name, APIServerState{currentCreatedTimestamp: pod.CreationTimestamp.Time})
+			continue
+		}
+
+		// the pod was recreated and the creation timestamp changed, update the last seen creation timestamps
+		// and start watching for termination events.
+		if pod.CreationTimestamp.Time != currentServerState.currentCreatedTimestamp {
+			if err := c.store.Update(pod.Name, APIServerState{currentCreatedTimestamp: pod.CreationTimestamp.Time}); err != nil {
+				return fmt.Errorf("unexpected sync error: %v", err)
+			}
+		}
```

Comment:
this whole loop would be a oneliner with a Upsert func in the store.

Reply:
i simplified that loop now